File: BHAVRP_Python/cujae/inf/citi/om/assignment/clustering/partitional/pam.py
import numpy as np
from typing import List
from .by_medoids import ByMedoids
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.input.location import Location
from ....problem.solution.solution import Solution
from ....problem.solution.cluster import Cluster

class PAM(ByMedoids):
    
    def __init__(self):
        super().__init__()
        self.list_id_elements: List[int]
        self.list_medoids: List[Depot]
        self.list_customers_to_assign: List[Customer]
        self.list_clusters: List[Cluster]
        
    def get_current_iteration(self) -> int:
        return self.current_iteration
    
    def to_clustering(self) -> Solution:
        self.initialize()
        self.assign()
        return self.finish()
        
    def initialize(self):
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_id_elements = self.generate_elements(self.list_customers_to_assign)
        self.list_clusters = self.initialize_clusters()
        self.list_medoids = []
        self.current_iteration = 0

    def assign(self):
        change: bool = True
        first: bool = True

        while change and self.current_iteration < self.count_max_iterations:
            if first:
                self.list_customers_to_assign = list(Problem.get_problem().get_customers())
                self.update_customer_to_assign()
                self.list_medoids = self.create_centroids()
                first = False
            else:
                self.update_clusters(self.list_clusters, self.list_id_elements)
                self.list_customers_to_assign = list(Problem.get_problem().get_customers())
                
            self.step_assignment(self.list_clusters, self.list_medoids)
            self.old_medoids = self.replicate_depots(self.list_medoids)
            self.best_cost = self.calculate_cost(self.list_medoids)
            self.step_search_medoids(self.list_clusters, self.list_medoids)

            change = self.verify_medoids(self.old_medoids, self.list_medoids)
            if change and (self.current_iteration + 1) != self.count_max_iterations:
                self.list_id_elements = self.get_id_medoids(self.list_medoids)
                self.clean_clusters(self.list_clusters)

            self.current_iteration += 1
            print(f"ITERACIÓN: {self.current_iteration}")
            
    def finish(self) -> Solution:
        solution = Solution()    
        
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                solution.get_total_unassigned_items().append(customer.get_id_customer())

        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    solution.get_clusters().append(cluster)
                    
        OSRMService.clear_distance_cache()
        
        return solution
    
    # Método que realiza la búsqueda de mejores medoides en cada clúster evaluando diferentes candidatos.
    def step_search_medoids(
        self, 
        clusters: List[Cluster], 
        medoids: List[Depot]
    ):
        current_cost = 0.0
        
        old_medoids: List[Depot] = self.replicate_depots(medoids)
        
        print("--------------------------------------------------------------------")
        print(f"PROCESO DE BÚSQUEDA")
        
        for i in range(len(clusters)):
            best_loc_medoid = Location(medoids[i].get_location_depot().get_axis_x(), medoids[i].get_location_depot().get_axis_y())
            
            print("--------------------------------------------------")
            print(f"MEJOR MEDOIDE ID: {medoids[i].get_id_depot()}")
            print(f"MEJOR MEDOIDE LOCATION X: {best_loc_medoid.get_axis_x()}")
            print(f"MEJOR MEDOIDE LOCATION Y: {best_loc_medoid.get_axis_y()}")
            print("--------------------------------------------------")
            
            for j in range(1, len(clusters[i].get_items_of_cluster())):
                new_id_medoid = clusters[i].get_items_of_cluster()[j]
                new_medoid = Customer()

                new_medoid.set_id_customer(Problem.get_problem().get_customer_by_id_customer(new_id_medoid).get_id_customer())
                new_medoid.set_request_customer(Problem.get_problem().get_customer_by_id_customer(new_id_medoid).get_request_customer())
                
                location = Location()
                location.set_axis_x(Problem.get_problem().get_customer_by_id_customer(new_id_medoid).get_location_customer().get_axis_x())
                location.set_axis_y(Problem.get_problem().get_customer_by_id_customer(new_id_medoid).get_location_customer().get_axis_y())
                new_medoid.set_location_customer(location)
                
                medoids[i].set_id_depot(new_id_medoid)
                medoids[i].set_location_depot(new_medoid.get_location_customer())
                
                print(f"ID DEL NUEVO MEDOIDE: {new_id_medoid}")
                print(f"X DEL NUEVO MEDOIDE: {new_medoid.get_location_customer().get_axis_x()}")
                print(f"Y DEL NUEVO MEDOIDE: {new_medoid.get_location_customer().get_axis_y()}")
                print("--------------------------------------------------")
                print("LISTA DE MEDOIDES")
                print(f"ID: {medoids[i].get_id_depot()}")
                print(f"X: {medoids[i].get_location_depot().get_axis_x()}")
                print(f"Y: {medoids[i].get_location_depot().get_axis_y()}")
                
                print("LISTA DE ANTERIORES MEDOIDES")
                print(f"ID: {old_medoids[i].get_id_depot()}")
                print(f"X: {old_medoids[i].get_location_depot().get_axis_x()}")
                print(f"Y: {old_medoids[i].get_location_depot().get_axis_y()}")
                print("--------------------------------------------------")
                
                current_cost = self.calculate_cost(medoids)
                
                print("---------------------------------------------")
                print(f"ACTUAL COSTO TOTAL: {current_cost}")
                print("---------------------------------------------")
                
                if current_cost < self.best_cost:
                    self.best_cost = current_cost
                    best_loc_medoid = medoids[i].get_location_depot()
                    
                    print(f"NUEVO MEJOR COSTO TOTAL: {self.best_cost}")
                    print(f"NUEVO MEDOIDE ID: {medoids[i].get_id_depot()}")
                    print(f"NUEVO MEDOIDE LOCATION X: {best_loc_medoid.get_axis_x()}")
                    print(f"NUEVO MEDOIDE LOCATION Y: {best_loc_medoid.get_axis_y()}")
                    print("---------------------------------------------")
                    
                    old_medoids[i].set_id_depot(medoids[i].get_id_depot())
                    old_medoids[i].get_location_depot().set_axis_x(medoids[i].get_location_depot().get_axis_x())
                    old_medoids[i].get_location_depot().set_axis_y(medoids[i].get_location_depot().get_axis_y())
                else:
                    medoids[i].set_id_depot(old_medoids[i].get_id_depot())
                    medoids[i].get_location_depot().set_axis_x(old_medoids[i].get_location_depot().get_axis_x())
                    medoids[i].get_location_depot().set_axis_y(old_medoids[i].get_location_depot().get_axis_y())
                
                print(f"ID MEDOIDE: {medoids[i].get_id_depot()}")
                print(f"LISTA DE MEDOIDES X: {medoids[i].get_location_depot().get_axis_x()}")
                print(f"LISTA DE MEDOIDES Y: {medoids[i].get_location_depot().get_axis_y()}")
                print("---------------------------------------------")
            
            medoids[i].set_location_depot(best_loc_medoid)
    
    # Método que calcula el costo total de los clústeres considerando los depósitos como medoides.
    def calculate_cost(
        self, 
        medoids: List[Depot]
    ) -> float:
        cost = 0.0
        
        print("-------------------------------------------------------------------------------")
        print("CÁLCULO DEL MEJOR COSTO")
        
        cost_matrix: np.ndarray = self.initialize_cost_matrix(Problem.get_problem().get_customers(), medoids, self.distance_type)

        for i, cluster in enumerate(self.list_clusters):
            pos_depot = Problem.get_problem().get_pos_element(medoids[i].get_id_depot())
            if pos_depot >= len(medoids):
                pos_depot = pos_depot % len(medoids)
            list_id_customers = list(cluster.get_items_of_cluster())
            
            print("-------------------------------------------------------------------------------")
            print(f"ID MEDOIDE: {medoids[i].get_id_depot()}")
            print(f"POSICIÓN DEL MEDOIDE: {pos_depot}")
            print(f"CLIENTES ASIGNADOS AL MEDOIDE: {list_id_customers}")
            print("-------------------------------------------------------------------------------")

            for customer_id in list_id_customers:
                pos_customer = Problem.get_problem().get_pos_element(customer_id)

                if pos_depot == pos_customer:
                    cost += 0.0
                    actual_cost = 0.0
                else:
                    actual_cost = cost_matrix[pos_depot, pos_customer]
                    cost += actual_cost
                
                print(f"ID CLIENTE: {customer_id}")
                print(f"POSICIÓN DEL CLIENTE: {pos_customer}")
                print(f"COSTO: {actual_cost}")
                print("-------------------------------------------------------------------------------")
                print(f"COSTO ACUMULADO: {cost}")
                print("-------------------------------------------------------------------------------")
        
        print(f"MEJOR COSTO TOTAL: {cost}")
        print("-------------------------------------------------------------------------------")
        
        return cost

File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/assignment/clustering/partitional/PAM.java
package cujae.inf.ic.om.assignment.clustering.partitional;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Depot;
import cujae.inf.ic.om.problem.input.Location;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.Cluster;
import cujae.inf.ic.om.problem.output.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.matrix.NumericMatrix;

public class PAM extends ByMedoids {
	private ArrayList<Integer> listIDElements;
	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;
	
	public PAM() {
		super();
	}

	@Override
	public Solution toClustering() {
		initialize();
		assign();
		return finish();
	}
	
	@Override	
	public void initialize() {	
		listIDElements = generateElements(seedType, distanceType);
		listClusters = initializeClusters(listIDElements);		
		listCustomersToAssign = new ArrayList<Customer>();
	}	
		
	@Override	
	public void assign() {	
		ArrayList<Depot> listMedoids = new ArrayList<Depot>();
		
		boolean change = true;
		boolean first = true;
		
		while((change) && (currentIteration < countMaxIterations))
		{
			listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
			updateCustomerToAssign(listCustomersToAssign, listIDElements);
			
			if(first)
			{
				listMedoids = createCentroids(listIDElements);	
				first = false;
			}
			else
				updateClusters(listClusters, listIDElements);
			
			NumericMatrix costMatrix = initializeCostMatrix(listCustomersToAssign, listMedoids, distanceType);
			NumericMatrix costMatrixCopy = new NumericMatrix(costMatrix); 
				
			stepAssignment(listClusters, listCustomersToAssign, costMatrix);
			ArrayList<Depot> oldMedoids = replicateDepots(listMedoids);
			
			double bestCost = calculateCost(listClusters, costMatrixCopy, listMedoids);
			
			stepSearchMedoids(listClusters, listMedoids, costMatrixCopy, bestCost);
			change = verifyMedoids(oldMedoids, listMedoids); 
			 
			if((change) && (currentIteration + 1 != countMaxIterations))
			{
				listIDElements.clear();
				listIDElements = getIDMedoids(listMedoids);
				cleanClusters(listClusters);
			}
			
			currentIteration++;
			
			System.out.println("ITERACIÓN: " + currentIteration);
		}
	}	

	@Override
	public Solution finish() {
		Solution solution = new Solution();
		
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));
		
		OSRMService.clearDistanceCache();
		
		return solution;
	}
	
	private void stepSearchMedoids(ArrayList<Cluster> clusters, ArrayList<Depot> medoids, NumericMatrix costMatrix, double bestCost){
		double currentCost = 0.0;
		
		ArrayList<Depot> oldMedoids = replicateDepots(medoids);
		
		System.out.println("--------------------------------------------------------------------");
		System.out.println("PROCESO DE BÚSQUEDA");
		
		for(int i = 0; i < clusters.size(); i++) 
		{	
			Location bestLocMedoid = new Location(medoids.get(i).getLocationDepot().getAxisX(), medoids.get(i).getLocationDepot().getAxisY());
			
			System.out.println("--------------------------------------------------");
			System.out.println("MEJOR MEDOIDE ID: " + medoids.get(i).getIDDepot());
			System.out.println("MEJOR MEDOIDE LOCATION X: " + bestLocMedoid.getAxisX());
			System.out.println("MEJOR MEDOIDE LOCATION Y: " + bestLocMedoid.getAxisY());
			System.out.println("--------------------------------------------------");
			
			for(int j = 1; j < clusters.get(i).getItemsOfCluster().size(); j++) 
			{
				int newIDMedoid = clusters.get(i).getItemsOfCluster().get(j);
				Customer newMedoid = new Customer();
				
				newMedoid.setIDCustomer(Problem.getProblem().getCustomerByIDCustomer(newIDMedoid).getIDCustomer());
				newMedoid.setRequestCustomer(Problem.getProblem().getCustomerByIDCustomer(newIDMedoid).getRequestCustomer());	
				
				Location location = new Location();
				location.setAxisX(Problem.getProblem().getCustomerByIDCustomer(newIDMedoid).getLocationCustomer().getAxisX());
				location.setAxisY(Problem.getProblem().getCustomerByIDCustomer(newIDMedoid).getLocationCustomer().getAxisY());
				newMedoid.setLocationCustomer(location);
				
				medoids.get(i).setIDDepot(newIDMedoid);
				medoids.get(i).setLocationDepot(newMedoid.getLocationCustomer());
				
				System.out.println("ID DEL NUEVO MEDOIDE: " + newIDMedoid);
				System.out.println("X DEL NUEVO MEDOIDE: " + newMedoid.getLocationCustomer().getAxisX());
				System.out.println("Y DEL NUEVO MEDOIDE: " + newMedoid.getLocationCustomer().getAxisY());
				
				System.out.println("--------------------------------------------------");
				System.out.println("LISTA DE MEDOIDES");
				System.out.println("ID: " + medoids.get(i).getIDDepot());
				System.out.println("X: " + medoids.get(i).getLocationDepot().getAxisX());
				System.out.println("Y: " + medoids.get(i).getLocationDepot().getAxisY());					
			
				System.out.println("LISTA DE ANTERIORES MEDOIDES");
				System.out.println("ID: " + oldMedoids.get(i).getIDDepot());
				System.out.println("X: " + oldMedoids.get(i).getLocationDepot().getAxisX());
				System.out.println("Y: " + oldMedoids.get(i).getLocationDepot().getAxisY());
				System.out.println("--------------------------------------------------");
								
				currentCost = calculateCost(clusters, costMatrix, medoids);
				
				System.out.println("---------------------------------------------");
				System.out.println("ACTUAL COSTO TOTAL: " + currentCost);
				System.out.println("---------------------------------------------");
				
				if(currentCost < bestCost) 
				{
					bestCost = currentCost;	
					bestLocMedoid = medoids.get(i).getLocationDepot();
					
					System.out.println("NUEVO MEJOR COSTO TOTAL: " + bestCost);
					System.out.println("NUEVO MEDOIDE ID: " + medoids.get(i).getIDDepot());
					System.out.println("NUEVO MEDOIDE LOCATION X: " + bestLocMedoid.getAxisX());
					System.out.println("NUEVO MEDOIDE LOCATION Y: " + bestLocMedoid.getAxisY());
					System.out.println("---------------------------------------------");
					
					oldMedoids.get(i).setIDDepot(medoids.get(i).getIDDepot());
					oldMedoids.get(i).getLocationDepot().setAxisX(medoids.get(i).getLocationDepot().getAxisX());
					oldMedoids.get(i).getLocationDepot().setAxisY(medoids.get(i).getLocationDepot().getAxisY());
				}
				else
				{
					medoids.get(i).setIDDepot(oldMedoids.get(i).getIDDepot());
					medoids.get(i).getLocationDepot().setAxisX(oldMedoids.get(i).getLocationDepot().getAxisX());					
					medoids.get(i).getLocationDepot().setAxisY(oldMedoids.get(i).getLocationDepot().getAxisY());	
				}
				
				System.out.println("ID MEDOIDE: " + medoids.get(i).getIDDepot());
				System.out.println("LISTA DE MEDOIDES X: " + medoids.get(i).getLocationDepot().getAxisX());
				System.out.println("LISTA DE MEDOIDES Y: " + medoids.get(i).getLocationDepot().getAxisY());
				System.out.println("---------------------------------------------");
			}
			
			medoids.get(i).setLocationDepot(bestLocMedoid);
		}
	}

	private double calculateCost(ArrayList<Cluster> clusters, NumericMatrix costMatrix, ArrayList<Depot> medoids) {
		double cost = 0.0;
		
		int idCustomer =  -1;
		int posCustomer = -1;
		int posDepot = -1;
		ArrayList<Integer> listIDCustomers = new ArrayList<Integer>();

		System.out.println("-------------------------------------------------------------------------------");
		System.out.println("CALCULO DEL MEJOR COSTO");
		
		for(int i = 0; i < clusters.size(); i++) 
		{			
			posDepot = Problem.getProblem().getPosElement(medoids.get(i).getIDDepot());
			listIDCustomers = clusters.get(i).getItemsOfCluster();
		
			System.out.println("-------------------------------------------------------------------------------");		
			System.out.println("ID MEDOIDE: " + medoids.get(i).getIDDepot());
			System.out.println("POSICIÓN DEL MEDOIDE: " + posDepot);
			System.out.println("CLIENTES ASIGNADOS AL MEDOIDE: " + listIDCustomers);
			System.out.println("-------------------------------------------------------------------------------");
			
			for(int j = 0; j < listIDCustomers.size(); j++) 
			{	
				idCustomer =  listIDCustomers.get(j);				
				posCustomer = Problem.getProblem().getPosElement(idCustomer);

				if(posDepot == posCustomer)
					cost += 0.0;
				else
					cost += costMatrix.getItem(posDepot, posCustomer);	
				
				System.out.println("ID CLIENTE: " + idCustomer);
				System.out.println("POSICIÓN DEL CLIENTE: " + posCustomer);
				System.out.println("COSTO: " + costMatrix.getItem(posDepot, posCustomer));
				System.out.println("-------------------------------------------------------------------------------");
				System.out.println("COSTO ACUMULADO: " + cost);	
				System.out.println("-------------------------------------------------------------------------------");
			}
		}
		System.out.println("MEJOR COSTO TOTAL: " + cost);	
		System.out.println("-------------------------------------------------------------------------------");
		
		return cost;
	}
}